Fixes the temp-biomodels github actions failure.

luciansmith opened the pull request:
The problem here stemmed from pyomexmeta failing to read the XML for some generated files.  We can work on that separately, but in the meantime, this wraps that call in a try/catch block that simply reports a failure to read the file instead of halting.

In addition, this fixes an issue I found on my own runs where UTF-8 characters were confusing the writer:  we now explicitly tell the writer that we are using utf-8.

luciansmith committed:
Fixes the temp-biomodels github actions failure.

The problem here stemmed from pyomexmeta failing to read the XML for some generated files.  We can work on that separately, but in the meantime, this wraps that call in a try/catch block that simply reports a failure to read the file instead of halting.

In addition, this fixes an issue I found on my own runs where UTF-8 characters were confusing the writer:  we now explicitly tell the writer that we are using utf-8.

jonrkarr commented on biosimulators_utils/omex_meta/utils.py:
```diff
@@ -51,12 +51,16 @@ def build_omex_meta_file_for_model(model_filename,
+    model = None
+    try:
+        model = editor.get_xml()
+    except:
```
There's lint error here
```
./biosimulators_utils/omex_meta/utils.py:57:5: E722 do not use bare 'except'
```

jonrkarr replied:
We're trying to be more specific rather than just catching everything. Can the specific type of exception be inserted here?

```python
except <TypeOfException`> as exception:
```

Ideally the exception message would be incorporated in line 61 below unless its uniformative.

luciansmith replied:
Makes sense.  This particular error was something low-level; I think it was an invalid read on my machine  I'll figure out what it was and revise.  

However, looking at the githubactions logs, it's an actual Segmentation Fault on linux, so that might be more problematic.  Fortunately, it only happened with biomodels 585, and there because it wrote out a series of metadataN.xml files which later got confused for SBML files (I think in the fix-entries code); this error is actually the result of reading in a metadata file and trying to parse it as SBML.  Obviously something that needs to be fixed on that end, but fortunately something I can fix on our end for the time being.

jonrkarr replied:
Segmentation faults are what I saw before. Previously I called pyomexmeta through a subprocess to capture this.

Its probably easiest to change the SBML than to wait for pyomexmeta.

jonrkarr commented on biosimulators_utils/omex_meta/utils.py:
```diff
@@ -51,12 +51,16 @@ def build_omex_meta_file_for_model(model_filename,
         raise RuntimeError('Model `{}` could not be read:\n  {}'.format(
             model_filename, errors.replace('\n', '\n  ')))
-    with open(model_filename, 'w') as file:
+    with open(model_filename, 'w', encoding="utf-8") as file:
```
I'm hesitant to assuming the encoding is utf-8. It would better to get the encoding from the model object if possible.

luciansmith replied:
That's a good point, though in this case wouldn't you need a method in pyomexmeta to tell you the encoding?  I suppose we could also use one of the encoding guessing python packages to guess for us.

I can say that it seems to me the COMBINE community has settled on UTF-8 as a standard, though there's certainly room for other groups to come in with something new.  But even then, I *think* a pyomex file would always still use the same encoding.

jonrkarr replied:
lxml, for example can report the encoding. Since pyomexmeta doesn't seem to be able to do this, lets let this go.

jonrkarr reviewed (changes requested):
RE: libomex, that's unfortunate. I did have more error handling before. I removed it after I pushed Ciaran to address issues with libOmex.

luciansmith committed:
Catch Exception instead of nothing.

On Linux, it's just a segfault, but a generic Exception on Windows does catch the low-level 'invalid read' the you get when libomexmeta tries to parse a non-SBML file as SBML.

jonrkarr commented:
Sounds like this should be closed because pyomexmeta fails with a segmentation fault rather than an exception that can be caught.

luciansmith commented:
It fails with a segmentation fault on linux, but with an Exception on Windows, so the fix does still help there, just not everywhere.  (I haven't investigated on Macs.)  The utf-8 encoding change definitely helps (though as we discussed, it'd be slightly nicer if we could ask pyomexmeta what the encoding should be).

I'll file an issue with pyomexmeta for the specific problem so it can be fixed more robustly going forward.

luciansmith commented:
Filed https://github.com/sys-bio/libOmexMeta/issues/139

jonrkarr committed:
feat: added handling for pyomexmeta segmentation faults, fixed error messages

jonrkarr reviewed (approved):
I added handling for segmentation faults
